server/manager/src/main/java/org/apache/accumulo/manager/compaction/queue/CompactionJobQueues.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.manager.compaction.queue;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.spi.compaction.CompactionExecutorId;
+import org.apache.accumulo.core.spi.compaction.CompactionJob;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CompactionJobQueues {
+
+  private static final Logger log = LoggerFactory.getLogger(CompactionJobQueues.class);
+
+  private final ConcurrentHashMap<CompactionExecutorId,CompactionJobPriorityQueue> priorityQueues =
+      new ConcurrentHashMap<>();
+
+  public void add(TabletMetadata tabletMetadata, Collection<CompactionJob> jobs) {
+    if (jobs.size() == 1) {
+      var executorId = jobs.iterator().next().getExecutor();
+      add(tabletMetadata, executorId, jobs);
+    } else {
+      jobs.stream().collect(Collectors.groupingBy(CompactionJob::getExecutor)).forEach(
+          ((executorId, compactionJobs) -> add(tabletMetadata, executorId, compactionJobs)));
+    }
+  }
+
+  public static class MetaJob {
+    private final CompactionJob job;
+
+    // the metadata from which the compaction job was derived
+    private final TabletMetadata tabletMetadata;
+
+    public MetaJob(CompactionJob job, TabletMetadata tabletMetadata) {
+      this.job = job;
+      this.tabletMetadata = tabletMetadata;
+    }
+
+    public CompactionJob getJob() {
+      return job;
+    }
+
+    public TabletMetadata getTabletMetadata() {
+      return tabletMetadata;
+    }
+  }
+
+  public MetaJob poll(CompactionExecutorId executorId) {
+    var prioQ = priorityQueues.get(executorId);
+    if (prioQ == null) {
+      return null;
+    }
+    MetaJob mj = prioQ.poll();
+
+    if(mj == null){
+      priorityQueues.computeIfPresent(executorId, (eid, pq)->{
```

Comment:
When priorityQueues are empty, it looks like this lambda doesn't remove the pq from the HashMap 
`priorityQueues`, it just sets the executorId key's mapped value to null.

I'm not super familiar with `computeIfPresent`, but shouldn't this involve a `priorityQueues.remove()` call?

Reply:
When null is returned the entry is removed from the map.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfPresent-K-java.util.function.BiFunction-


This is a TODO for myself that was prompted by this comment. I need to add some comments int the code about depending on specific sync behavior of concurrent hash map.